odoo/addons/base/models/res_users.py

```diff
@@ -843,7 +847,8 @@ def _login(cls, db, login, password, user_agent_env):
                     if not user:
                         raise AccessDenied()
                     user = user.with_user(user)
-                    user._check_credentials(password, user_agent_env)
+                    auth_info = user._check_credentials(credential, user_agent_env)
+                    auth_info['uid'] = user.id
```

Comment:
> What exactly do you mean by this? Could you elaborate a bit why we would need to assert that user(?) is an empty recordset?

He means that the `_check_credentials` method doesn't care with which user records you call it with.
You would pass another user than yours, or the whole set of users of the database, it would still make the check against `self.env.user`, not `self`.
https://github.com/odoo/odoo/blob/8be83ac3f94dfe85f501964fa6ed1f4a0b9498e9/odoo/addons/base/models/res_users.py#L431-L434

Hence, the confusion would be that a developer might think that `_check_credentials` is checked against the records you call the method with.
For instance, reading `self.env.ref('base.user_portal')._check_credentials(...)`, the developer would think it's the portal user credentials which are checked, while in reality it's the credentials of the user you are currently logged in with which are checked.

Adding the `@api.model` decorator on the method, and an assert in the beginning of the method checking `self` is an empty recordset, will make things clearer.

However, in my opinion, this is out of scope of this pull request. This should be done in a standalone task, cleaning/stricting the whole authentication framework. For which I already have a task planned.